File: backend/app/services/version_app.py
"""Que version del APK esta vigente y cual ya no se puede usar.

GeoMaps se reparte como APK directo, sin Play Store, asi que nadie empuja las
actualizaciones: la app pregunta y este modulo contesta.

## De donde sale la respuesta

De un manifiesto en el bucket, `app/version.json`, que escribe
`tools/publicar_apk.py` al publicar un APK. Vive en S3 y no en variables de
Vercel a proposito: publicar una version no puede exigir un redeploy del
backend, y el manifiesto y el APK tienen que cambiar juntos.

```json
{
  "version_code": 2,             // el +N del pubspec
  "version_nombre": "0.2.0",
  "publicada_en": "2026-09-22T15:00:00+00:00",
  "dias_gracia": 10,
  "minima": 1,                   // por debajo de esto, bloqueo inmediato
  "llave_apk": "app/geomaps-0.2.0+2.apk",
  "sha256": "...",
  "tamano_bytes": 52428800,
  "firma_sha256": "...",         // huella del certificado que firmo el APK
  "notas": "Mapas nuevos de ..."
}
```

## La regla

| Version del telefono | Estado |
|---|---|
| `>= version_code` | `al_dia` |
| `< minima` | `bloqueada` |
| `< version_code`, antes de `publicada_en + dias_gracia` | `desactualizada`: aviso, se sigue trabajando |
| `< version_code`, despues de ese plazo | `bloqueada` |

La app aplica exactamente la misma regla con el manifiesto que guardo la
ultima vez que tuvo red, asi que el bloqueo llega aunque el telefono no vuelva
a ver senal. Y el backend la aplica en cada request (ver `main.py`): una app
vieja que se salte el chequeo igual recibe 426 al sincronizar, y no puede subir
datos con un formato que ya no existe.
"""

from __future__ import annotations

import asyncio
import json
import logging
import time
from datetime import datetime, timedelta, timezone
from typing import Literal

from pydantic import BaseModel

from ..config import Settings
from . import almacenamiento

log = logging.getLogger(__name__)

Estado = Literal["al_dia", "desactualizada", "bloqueada"]

# Cuanto se confia en el manifiesto leido antes de volver a S3. Lo lee el
# middleware en cada request: sin cache, cada llamada de la app pagaria una
# lectura al bucket. Un minuto es lo que tarda en verse una version recien
# publicada, y nadie la esta esperando con el cronometro.
TTL_CACHE_SEGUNDOS = 60


class Manifiesto(BaseModel):
    """La version vigente, tal como la dejo `tools/publicar_apk.py`."""

    version_code: int
    version_nombre: str
    publicada_en: datetime
    dias_gracia: int = 10
    minima: int = 0
    llave_apk: str
    sha256: str
    tamano_bytes: int = 0
    firma_sha256: str = ""
    notas: str = ""

    @property
    def bloquea_en(self) -> datetime:
        """Desde cuando las versiones anteriores a esta dejan de servir."""
        publicada = self.publicada_en
        if publicada.tzinfo is None:
            publicada = publicada.replace(tzinfo=timezone.utc)
        return publicada + timedelta(days=self.dias_gracia)


def evaluar(version_code: int, manifiesto: Manifiesto, ahora: datetime) -> Estado:
    """La regla entera. La app tiene una copia identica en `version_app.dart`."""
    if version_code >= manifiesto.version_code:
        return "al_dia"
    if version_code < manifiesto.minima or ahora >= manifiesto.bloquea_en:
        return "bloqueada"
    return "desactualizada"


def version_de_header(valor: str | None) -> int | None:
    """Lee `X-App-Version`. Un valor que no es un entero se trata como ausente:
    no hay que tumbar un request por una cabecera rara."""
    if not valor:
        return None
    try:
        return int(valor.strip())
    except ValueError:
        return None


# --- Lectura del manifiesto con cache --------------------------------------

_cache: tuple[float, Manifiesto | None] | None = None


def _leer_de_s3(settings: Settings) -> Manifiesto | None:
    crudo = almacenamiento.leer(settings, settings.s3_llave_manifiesto_app)
    if crudo is None:
        return None
    return Manifiesto.model_validate(json.loads(crudo))


async def manifiesto_vigente(settings: Settings) -> Manifiesto | None:
    """El manifiesto publicado, o None si todavia no se publico ninguno.

    Si S3 falla, devuelve None y se cachea igual: sin manifiesto no se bloquea
    a nadie. Es la falla correcta — un bucket caido no puede dejar a todo el
    equipo sin app, y lo peor que pasa es que una version vieja trabaje un
    minuto mas.
    """
    global _cache
    ahora = time.monotonic()
    if _cache is not None and ahora - _cache[0] < TTL_CACHE_SEGUNDOS:
        return _cache[1]

    try:
        manifiesto = await asyncio.to_thread(_leer_de_s3, settings)
    except Exception:  # noqa: BLE001
        log.exception("No se pudo leer el manifiesto de version")
        manifiesto = None

    _cache = (ahora, manifiesto)
    return manifiesto


def olvidar_cache() -> None:
    """Para las pruebas."""
    global _cache
    _cache = None

File: app/lib/core/version_app.dart
/// Que version del APK esta vigente, y si la de este telefono todavia sirve.
///
/// La app se reparte como APK directo, sin Play Store: nadie empuja las
/// actualizaciones. La app le pregunta al backend (`GET /v1/version`), guarda
/// la respuesta, y con esa copia decide sola aunque no vuelva a tener senal.
///
/// La regla es la misma que aplica el backend en `services/version_app.py`, y
/// tiene que seguir siendolo: si la app cree que puede trabajar y el backend
/// no, la persona trabaja una semana y despues la sincronizacion le rebota.
///
/// | Version del telefono | Estado |
/// |---|---|
/// | `>= versionCode` | al dia |
/// | `< minima` | bloqueada, sin plazo |
/// | `< versionCode`, antes de `bloqueaEn` | desactualizada: aviso y sigue |
/// | `< versionCode`, desde `bloqueaEn` | bloqueada |
///
/// `bloqueaEn` es la fecha de publicacion mas los dias de gracia (10), y la
/// calcula el backend.
library;

enum EstadoVersion { alDia, desactualizada, bloqueada }

/// La version que corre en este telefono. Sale de `package_info_plus`, que la
/// lee del APK: es el `version:` del pubspec, `nombre+code`.
class VersionLocal {
  const VersionLocal({required this.code, required this.nombre});

  final int code;
  final String nombre;
}

/// La version vigente segun el ultimo manifiesto que se vio con red.
class VersionPublicada {
  const VersionPublicada({
    required this.versionCode,
    required this.versionNombre,
    required this.minima,
    required this.bloqueaEn,
    this.notas = '',
    this.tamanoBytes = 0,
    this.sha256,
    this.apkUrl,
  });

  factory VersionPublicada.deJson(Map<String, dynamic> j) => VersionPublicada(
    versionCode: j['version_code'] as int,
    versionNombre: j['version_nombre'] as String? ?? '',
    minima: j['minima'] as int? ?? 0,
    bloqueaEn: DateTime.parse(j['bloquea_en'] as String),
    notas: j['notas'] as String? ?? '',
    tamanoBytes: j['tamano_bytes'] as int? ?? 0,
    sha256: j['sha256'] as String?,
    apkUrl: j['apk_url'] as String?,
  );

  final int versionCode;
  final String versionNombre;
  final int minima;
  final DateTime bloqueaEn;
  final String notas;
  final int tamanoBytes;
  final String? sha256;

  /// Prefirmada, vence en un dia. Por eso antes de descargar se vuelve a
  /// consultar el backend en vez de usar la guardada.
  final String? apkUrl;

  Map<String, dynamic> aJson() => {
    'version_code': versionCode,
    'version_nombre': versionNombre,
    'minima': minima,
    'bloquea_en': bloqueaEn.toIso8601String(),
    'notas': notas,
    'tamano_bytes': tamanoBytes,
    'sha256': sha256,
    'apk_url': apkUrl,
  };

  /// La regla entera. Copia de `evaluar` en el backend.
  EstadoVersion evaluar(int local, DateTime ahora) {
    if (local >= versionCode) return EstadoVersion.alDia;
    if (local < minima || !ahora.isBefore(bloqueaEn)) {
      return EstadoVersion.bloqueada;
    }
    return EstadoVersion.desactualizada;
  }

  /// Dias que quedan antes del bloqueo, redondeando para arriba: con 3 horas
  /// por delante, "0 dias" haria creer que ya esta bloqueada.
  int diasParaBloqueo(DateTime ahora) {
    final falta = bloqueaEn.difference(ahora);
    if (falta.isNegative) return 0;
    return (falta.inMinutes / (60 * 24)).ceil();
  }

  /// "52 MB", para que la persona decida si lo baja con datos moviles.
  String get tamanoLegible {
    if (tamanoBytes <= 0) return '';
    return '${(tamanoBytes / (1024 * 1024)).toStringAsFixed(0)} MB';
  }
}

/// La hora contra la que se evalua.
///
/// Es la mayor entre el reloj del telefono y la ultima hora confiable que se
/// vio (la del servidor, o la del propio telefono en un chequeo anterior).
/// Atrasar el reloj a mano no esquiva el bloqueo: la hora nunca va para atras.
DateTime horaEfectiva(DateTime reloj, DateTime? ultimaConfiable) {
  if (ultimaConfiable == null || reloj.isAfter(ultimaConfiable)) return reloj;
  return ultimaConfiable;
}
